solver/src/solver/solver_utils.rs

```diff
@@ -0,0 +1,94 @@
+use anyhow::{ensure, Result};
+use contracts::{GPv2Settlement, ERC20};
+use ethcontract::H160;
+use ethcontract::U256;
+use serde::{
+    de::{Deserializer, Error as _},
+    Deserialize,
+};
+use web3::types::Bytes;
+
+use std::{
+    borrow::Cow,
+    fmt::{self, Display, Formatter},
+};
+
+// Helper trait to mock the smart contract interaction
+#[cfg_attr(test, mockall::automock)]
+#[async_trait::async_trait]
+pub trait AllowanceFetching: Send + Sync {
+    async fn existing_allowance(&self, token: H160, spender: H160) -> Result<U256>;
+}
+
+#[async_trait::async_trait]
+impl AllowanceFetching for GPv2Settlement {
+    async fn existing_allowance(&self, token: H160, spender: H160) -> Result<U256> {
+        let token_contract = ERC20::at(&self.raw_instance().web3(), token);
+        Ok(token_contract
+            .allowance(self.address(), spender)
+            .call()
+            .await?)
+    }
+}
+
+/// A slippage amount.
+#[derive(Clone, Copy, Debug, PartialEq, PartialOrd, Default)]
+pub struct Slippage(pub f64);
+
+impl Slippage {
+    /// Creates a slippage amount from the specified percentage.
+    pub fn percentage(amount: f64) -> Result<Self> {
+        // 1Inch API only accepts a slippage from 0 to 50.
+        ensure!(
+            (0. ..=50.).contains(&amount),
+            "slippage outside of [0%, 50%] range"
+        );
+        Ok(Slippage(amount))
+    }
+
+    /// Creates a slippage amount from the specified basis points.
+    pub fn basis_points(bps: u16) -> Result<Self> {
+        let percent = (bps as f64) / 100.;
+        Slippage::percentage(percent)
+    }
+}
+
+impl Display for Slippage {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+pub fn deserialize_decimal_f64<'de, D>(deserializer: D) -> Result<f64, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let decimal_str = Cow::<str>::deserialize(deserializer)?;
```

Comment:
Why do we need Cow here?

Reply:
Coincidence of wants of cource :stuck_out_tongue:.

Joking aside, I had in the past had issues with `visit_str` vs `visit_borrowed_str` vs `visit_string` and `deserialize_str` vs `deserialize_string` with deserialization with unexpected results, and using `Cow` (which can be either a `Borrowed` or `Owned` string) solved the issue.

However, looking at the documentation, it seems to suggest that this shouldn't be an issue - so maybe we should just stick with `str` and see if there is an issue?

Reply:
@josojo can you comment why cow is needed or address this suggestion?

Reply:
I just copied this line from the existing code in 1inch. Since I don't know in which issues nlordell was running, I don't know whether they will reoccur. Hence, I would just not change it, as it runs currently.

Reply:
So I did some investigation because I was curious and:
```rust
println!("{:?}", <&str>::deserialize(json!("asdf"))); // Err(Error("invalid type: string \"asdf\", expected a borrowed string", line: 0, column: 0))
```

This is because:
> The reason serde can't deserialize into a borrowed string is that it needs to modify (unescape) the encoded json, which requires ownership.
>
> That said, it appears you can get around this by using a `Cow<'a, str>` allowing deserialization to borrowed or owned memory depending on the string itself.